### PR TITLE
Fix GL Push Constant Layout

### LINKED ISSUE #3375: push_constant_input test fails on GL backend
**Description**
The test was implicitly skipped before as the GL backend didn't meet the `max_push_constant_size` limits. After #3374 increased that, the test then ran and fails with the following panic:

```
thread 'shader::struct_layout::push_constant_input' panicked at 'No uniform for push constant', wgpu-hal/src/gles/command.rs:674:17
stack backtrace:
   0: rust_begin_unwind
             at /rustc/69f9c33d71c871fc16ac445211281c6e7a340943/library/std/src/panicking.rs:575:5
   1: core::panicking::panic_fmt
             at /rustc/69f9c33d71c871fc16ac445211281c6e7a340943/library/core/src/panicking.rs:65:14
   2: wgpu_hal::gles::command::<impl wgpu_hal::CommandEncoder<wgpu_hal::gles::Api> for wgpu_hal::gles::CommandEncoder>::set_push_constants
             at /home/dinnerbone/workspace/personal/wgpu/wgpu-hal/src/gles/command.rs:674:17
   3: wgpu_core::command::compute::<impl wgpu_core::hub::Global<G>>::command_encoder_run_compute_pass_impl::{{closure}}
             at /home/dinnerbone/workspace/personal/wgpu/wgpu-core/src/command/compute.rs:525:37
   4: wgpu_core::command::push_constant_clear
             at /home/dinnerbone/workspace/personal/wgpu/wgpu-core/src/command/mod.rs:458:9
   5: wgpu_core::command::compute::<impl wgpu_core::hub::Global<G>>::command_encoder_run_compute_pass_impl
             at /home/dinnerbone/workspace/personal/wgpu/wgpu-core/src/command/compute.rs:521:29
   6: wgpu_core::command::compute::<impl wgpu_core::hub::Global<G>>::command_encoder_run_compute_pass
             at /home/dinnerbone/workspace/personal/wgpu/wgpu-core/src/command/compute.rs:323:9
   7: <wgpu::backend::direct::Context as wgpu::context::Context>::command_encoder_end_compute_pass
             at ./src/backend/direct.rs:1815:29
   8: <T as wgpu::context::DynContext>::command_encoder_end_compute_pass
             at ./src/context.rs:2609:9
   9: <wgpu::ComputePass as core::ops::drop::Drop>::drop
             at ./src/lib.rs:3566:13
  10: core::ptr::drop_in_place<wgpu::ComputePass>
             at /rustc/69f9c33d71c871fc16ac445211281c6e7a340943/library/core/src/ptr/mod.rs:490:1
  11: core::mem::drop
             at /rustc/69f9c33d71c871fc16ac445211281c6e7a340943/library/core/src/mem/mod.rs:987:24
  12: wgpu_tests::shader::shader_input_output_test
             at ./tests/shader/mod.rs:337:9
  13: wgpu_tests::shader::struct_layout::push_constant_input::{{closure}}
             at ./tests/shader/struct_layout.rs:226:13
  14: wgpu_tests::common::initialize_test::{{closure}}
             at ./tests/common/mod.rs:295:53
  15: <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
             at /rustc/69f9c33d71c871fc16ac445211281c6e7a340943/library/core/src/panic/unwind_safe.rs:271:9
  16: std::panicking::try::do_call
             at /rustc/69f9c33d71c871fc16ac445211281c6e7a340943/library/std/src/panicking.rs:483:40
  17: __rust_try
  18: std::panicking::try
             at /rustc/69f9c33d71c871fc16ac445211281c6e7a340943/library/std/src/panicking.rs:447:19
  19: std::panic::catch_unwind
             at /rustc/69f9c33d71c871fc16ac445211281c6e7a340943/library/std/src/panic.rs:137:14
  20: wgpu_tests::common::initialize_test
             at ./tests/common/mod.rs:295:20
  21: wgpu_tests::shader::struct_layout::push_constant_input
             at ./tests/shader/struct_layout.rs:217:5
  22: wgpu_tests::shader::struct_layout::push_constant_input::{{closure}}
             at ./tests/shader/struct_layout.rs:216:1
  23: core::ops::function::FnOnce::call_once
             at /rustc/69f9c33d71c871fc16ac445211281c6e7a340943/library/core/src/ops/function.rs:251:5
  24: core::ops::function::FnOnce::call_once
             at /rustc/69f9c33d71c871fc16ac445211281c6e7a340943/library/core/src/ops/function.rs:251:5
```

Upon investigating it, the test creates shaders like:
```wgsl
struct CustomStruct { member: vec2<f32>, }
var<push_constant> input: CustomStruct;
```
whilst also setting the render pipeline to have a large push constant range:
```rust
let pll = ctx
        .device
        .create_pipeline_layout(&PipelineLayoutDescriptor {
            label: None,
            bind_group_layouts: &[&bgl],
            push_constant_ranges: match storage_type {
                InputStorageType::PushConstant => &[PushConstantRange {
                    stages: ShaderStages::COMPUTE,
                    range: 0..MAX_BUFFER_SIZE as u32,
                }],
                _ => &[],
            },
        });
```

This panics when the GL backend tries to write MAX_BUFFER_SIZE (128) bytes of data to a uniform that can only hold 8.

I'm not sure what the correct thing to do here is, why don't the other backends error too?

**Repro steps**
- Enable the test for GL again in `push_constant_input`
- run `WGPU_BACKEND=gl nextest run -- push_constant_input`

**Expected vs observed behavior**
Either other backends should error too, or the GL backend should silently ignore out of range inputs.

**Extra materials**
None

**Platform**
GL backend only.

### LINKED ISSUE #3822: DX12 Push Constant out of bound
**Description**
I think I found an issue with the current implementation of DX12 push constants. Currently if I try to use anything bigger than 64 bytes `wgpu` panics but from the docs and also the limits I should be able to use more.  I think the problem is how the offset is currently used/calculated on this line. 

https://github.com/gfx-rs/wgpu/blob/e6314c407048ae20a49c6cba0730949d5bb1a7e8/wgpu-hal/src/dx12/command.rs#L890

The offset is the one originally passed in from the renderpass and that's in bytes but root constants are expected to use 4 byte offsets. 

I think a fix for the issue could look something like this: 
```rust 
self.pass.constant_data[((offset / 4) as usize)..((offset / 4) as usize + data.len())]
          .copy_from_slice(data);
```

**Repro steps**
This is the project where I able to produce the issue. 
https://github.com/abdo643-HULK/Bachelor-Thesis/tree/main/source-code/app

**Expected vs observed behavior**
Expected: Just pass the offset in bytes like for the Vulkan backend.
Observed: DX12 uses incorrect size/offset.

**Extra materials**
image

**Platform**
OS: Windows 11 (build 22621.1778)
wgpu: 0.16.1

This is my first real bug report I hope this is helpful.

### PR DESCRIPTION
**Pull request by cwfitzgerald**
**Checklist**

- [x] Run `cargo clippy`.
- [x] Run `cargo clippy --target wasm32-unknown-unknown` if applicable.
- [x] Add change to CHANGELOG.md. See simple instructions inside file.

**Connections**

Closes https://github.com/gfx-rs/wgpu/issues/3375
Closes https://github.com/gfx-rs/wgpu/issues/3822

**Description**

We had multiple issues with the push constant layout algorithm. OpenGL's DCE of uniforms easily threw it off.

This completely reworks how we assert the locations of things, and improves our ability to upload the uniforms.

Depends on https://github.com/grovesNL/glow/pull/265

**Testing**

It is tested!

### CODE REVIEW

File: wgpu-hal/src/gles/mod.rs [hunk 1]
```diff
@@ -108,6 +108,8 @@ const MAX_SAMPLERS: usize = 16;
 const MAX_VERTEX_ATTRIBUTES: usize = 16;
 const ZERO_BUFFER_SIZE: usize = 256 << 10;
 const MAX_PUSH_CONSTANTS: usize = 64;
+// We have to account for each push constant may need to be set for every shader.
+const MAX_PUSH_CONSTANT_COMMANDS: usize = MAX_PUSH_CONSTANTS + crate::MAX_CONCURRENT_SHADER_STAGES;
```
Comment (cwfitzgerald) [inline wgpu-hal/src/gles/mod.rs]: Should be *

File: wgpu-hal/src/gles/conv.rs [hunk 2]
```diff
@@ -417,104 +417,74 @@ pub(super) fn map_storage_access(access: wgt::StorageTextureAccess) -> u32 {
...
-        | glow::IMAGE_2D_MULTISAMPLE
-        | glow::IMAGE_2D_MULTISAMPLE_ARRAY
-        | glow::IMAGE_2D_RECT
-        | glow::IMAGE_3D
-        | glow::IMAGE_CUBE
-        | glow::IMAGE_CUBE_MAP_ARRAY => true,
-        _ => false,
-    }
-}
+pub(super) fn map_naga_uniform_type(ty: &naga::TypeInner) -> u32 {
```
Comment (teoxoy) [inline wgpu-hal/src/gles/conv.rs]: I see this is only used for `uniform_byte_size` and the match statement.
Can we keep using `TypeInner` and its `size` method instead?

File: naga/src/back/glsl/mod.rs [hunk 3]
```diff
@@ -328,6 +330,50 @@ pub struct TextureMapping {
...
+    /// uniform PushConstants _push_constant_binding_cs;
+    /// ```
+    ///
+    /// ```text
+    /// - _push_constant_binding_cs.inner.value
+    /// - _push_constant_binding_cs.array[0]
+    /// - _push_constant_binding_cs.array[1]
+    /// ```
+    ///
+    pub name: String,
```
Comment (teoxoy) [inline naga/src/back/glsl/mod.rs]: It's a bit odd that this is called `name`. Maybe `access_path`? `location_name`?

File: wgpu-hal/src/gles/command.rs [hunk 4]
```diff
@@ -30,10 +29,41 @@ pub(super) struct State {
     instance_vbuf_mask: usize,
     dirty_vbuf_mask: usize,
     active_first_instance: u32,
-    push_offset_to_uniform: ArrayVec<super::UniformDesc, { super::MAX_PUSH_CONSTANTS }>,
+    uniform_descs: ArrayVec<super::UniformDesc, { super::MAX_PUSH_CONSTANT_COMMANDS }>,
```
Comment (teoxoy) [inline wgpu-hal/src/gles/command.rs]:
```suggestion
    push_constant_descs: ArrayVec<super::UniformDesc, { super::MAX_PUSH_CONSTANT_COMMANDS }>,
```

File: wgpu-hal/src/gles/mod.rs [hunk 5]
```diff
@@ -510,7 +512,7 @@ type SamplerBindMap = [Option<u8>; MAX_TEXTURE_SLOTS];
 struct PipelineInner {
     program: glow::Program,
     sampler_map: SamplerBindMap,
-    uniforms: [UniformDesc; MAX_PUSH_CONSTANTS],
+    uniforms: ArrayVec<UniformDesc, MAX_PUSH_CONSTANT_COMMANDS>,
```
Comment (teoxoy) [inline wgpu-hal/src/gles/mod.rs]:
```suggestion
    push_constant_descs: ArrayVec<UniformDesc, MAX_PUSH_CONSTANT_COMMANDS>,
```

**Review by teoxoy — CHANGES_REQUESTED** [review]
Good stuff! Left a few suggestions.

File: wgpu-hal/src/gles/conv.rs (hunk 2 above)
Comment (cwfitzgerald) [reply wgpu-hal/src/gles/conv.rs]: Nice, didn't know that existed!

File: naga/src/back/glsl/mod.rs (hunk 3 above)
Comment (cwfitzgerald) [reply naga/src/back/glsl/mod.rs]: I like access_path!

**Review by teoxoy — APPROVED** [review]

**Comment by teoxoy** [comment]
Can't merge, there seem to be some conflicts now.

**Comment by cwfitzgerald** [comment]
Weird, I could merge from my phone

**Pull request merged**